### PR TITLE
[HttpKernel] Replace ArgumentValueResolverInterface by ValueResolverInterface

### LINKED ISSUE #17225: [HttpKernel] Replace ArgumentValueResolverInterface by ValueResolverInt…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#47363
| PR author(s) | @nicolas-grekas
| Merged in    | 6.2

### PR DESCRIPTION
**Pull request by wouterj**
Fixes #17225 

I've also updated the example using a different example (one we use frequently in our projects). The issue with the current example is that it re-implements something that was already built-in, requiring all sorts of workarounds.

cc @nicolas-grekas

### CODE REVIEW

File: controller/value_resolver.rst [hunk 1]
```diff
@@ -168,132 +168,88 @@ PSR-7 Objects Resolver:
...
     {
-        private $security;
-
-        public function __construct(Security $security)
-        {
-            $this->security = $security;
-        }
-
-        public function supports(Request $request, ArgumentMetadata $argument): bool
+        public function resolve(Request $request, ArgumentMetadata $argument): iterable
```
Comment (nicolas-grekas) [inline controller/value_resolver.rst]: let it return an array, that's more specific

**Review by nicolas-grekas — APPROVED** [review]

File: controller/value_resolver.rst [hunk 2]
```diff
@@ -168,132 +168,88 @@ PSR-7 Objects Resolver:
...
-    use App\Entity\User;
-    use Symfony\Bundle\SecurityBundle\Security\Security;
+    use App\IdentifierInterface;
     use Symfony\Component\HttpFoundation\Request;
-    use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+    use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
     use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
 
-    class UserValueResolver implements ArgumentValueResolverInterface
+    class UserValueResolver implements ValueResolverInterface
```
Comment (javiereguiluz) [inline controller/value_resolver.rst:217]: `class UserValueResolver` -> `class BookingIdValueResolver` ?

File: controller/value_resolver.rst [hunk 3]
```diff
@@ -168,132 +168,88 @@ PSR-7 Objects Resolver:
...
 
-Both methods get the ``Request`` object, which is the current request, and an
+This interface contains a ``resolve()`` method, which is called for each
+argument of the controller. It receives the current ``Request`` object and an
 :class:`Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadata`
-instance. This object contains all information retrieved from the method signature
-for the current argument.
+instance, which contains all information from the method signature. The
+method should return either an empty array (if it cannot resolve this
+argument) or an array with the resolved value(s).
```
Comment (javiereguiluz) [inline controller/value_resolver.rst:205]: Why does it return always an array? What happens when you return more than a value in the array? Which one is picked? Thanks.

Comment (wouterj) [reply controller/value_resolver.rst:205]: The array is used for variadic arguments. You'll get a logic exception if you return more than one item and the argument isn't variadic

Comment (javiereguiluz) [reply controller/value_resolver.rst:205]: Ah, it makes sense now. Thanks.

**Comment by javiereguiluz** [comment]
Thanks Wouter for this fantastic contribution!

While merging I did some tweaks as reported in the review: https://github.com/symfony/symfony-docs/commit/262babab34042927db6d15e38aa8dcfc4e06b8b8

**Pull request merged**